Sources/_StringProcessing/ByteCodeGen.swift

```diff
@@ -74,6 +74,77 @@ fileprivate extension Compiler.ByteCodeGen {
     }
   }
 
+  mutating func emitQuotedLiteral(_ s: String) {
+    if options.semanticLevel == .graphemeCluster {
+      if options.isCaseInsensitive {
+        // future work: if all ascii, emit matchBitset instructions with
+        // case insensitive bitsets
+
+        // TODO: buildCaseInsensitiveMatchSequence(c) or alternative
+        builder.buildConsume { input, bounds in
+          var iterator = s.makeIterator()
+          var currentIndex = bounds.lowerBound
+          while let ch = iterator.next() {
+            guard currentIndex < bounds.upperBound,
+                  ch.lowercased() == input[currentIndex].lowercased()
+            else { return nil }
+            input.formIndex(after: &currentIndex)
+          }
+          return currentIndex
+        }
+      } else {
+        if optimizationsEnabled && s.allSatisfy({char in char.isASCII}) {
+          for char in s.dropLast(1) {
+            // Note: only cr-lf is multiple scalars
+            for scalar in char.unicodeScalars {
+              builder.buildMatchScalar(scalar, boundaryCheck: false)
+            }
+          }
+          let lastChar = s.last!
+          for scalar in lastChar.unicodeScalars {
+            // Only boundary check if we are the last scalar in the last character
+            // to make sure that there isn't a combining scalar after the quoted literal
+            let boundaryCheck = scalar == lastChar.unicodeScalars.last!
+            builder.buildMatchScalar(scalar, boundaryCheck: boundaryCheck)
+          }
```

Comment:
I think it would be simpler to say something like

```
let lastIdx = s.unicodeScalars.indicies.last! // BTW, make sure we have a non-empty invariant
for idx in s.unicodeScalars.indices {
  ...
  ... if isGraphemeSemantic && idx == lastIdx {
    ...
  } else { 
    ...
  }
}
```